Mathlib/Tactic/Linter/DeprecateModule.lean

```diff
@@ -0,0 +1,138 @@
+is imported.
+
+The usage is as follows. Write
+```lean
+import B
+...
+import Z
+
+deprecate_module since yyyy-mm-dd
+```
+in module `A` with the expectation that `A` contains nothing else.
+This triggers the `deprecateModule` linter to notify every file with `import A`
+to instead import the *direct imports* of `A`, that is `B,...,Z`.
+-/
+
+open Lean Elab Command
+
+namespace Mathlib.Linter
+
+/--
+The `deprecateModule` linter emits a warning when a file that has been renamed or split
+is imported.
+-/
+register_option linter.deprecateModule : Bool := {
+  defValue := true
+  descr := "enable the deprecateModule linter"
+}
+
+/--
+Defines the `deprecateModuleExt` extension for adding a `HashSet` of pairs of
+* a module `Name` that has been deprecated and
+* an array of `Name`s of modules that should be imported instead
+
+to the environment.
+-/
+initialize deprecateModuleExt :
+    SimplePersistentEnvExtension (Name × Array Name) (Std.HashSet (Name × Array Name)) ←
+  registerSimplePersistentEnvExtension {
+    addImportedFn := fun as => as.foldl Std.HashSet.insertMany {}
+    addEntryFn := .insert
+  }
+
+/--
+`addModuleDeprecation` adds to the `deprecateModuleExt` extension the pair consisting of the
+current module name and the array of its direct imports.
+
+It ignores `Init`, since this is a special module that is expected to be imported by all files.
+-/
+def addModuleDeprecation {m : Type → Type} [Monad m] [MonadEnv m] [MonadQuotation m] : m Unit := do
+  modifyEnv (deprecateModuleExt.addEntry ·
+    (← getMainModule, (← getEnv).imports.filterMap fun i =>
+      if i.module == `Init then none else i.module))
+
+/--
+`deprecate_module since yyyy-mm-dd` deprecates the current module `A` in favour of its imports.
+This means that any file that imports `A` will get a notification on the `import A` line suggesting
+to instead import the *direct imports* of `A`.
+-/
+elab (name := deprecate_modules)
+    "deprecate_module " &"since" yyyy:num "-" mm:num "-" dd:num: command => do
+  if yyyy.getNat < 2025 then
+    throwErrorAt yyyy "The year should be at least 2025!"
+  if mm.getNat == 0 || 12 < mm.getNat || mm.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt mm "The month should be of the form 01, 02,..., 12!"
+  if dd.getNat == 0 || 31 < dd.getNat || dd.raw.getSubstring?.get!.toString.trim.length != 2 then
+    throwErrorAt dd "The day should be of the form 01, 02,..., 31!"
+  addModuleDeprecation
+  -- disable the linter, so that it does not complain in the file with the deprecation
+  elabCommand (← `(set_option linter.deprecateModule false))
+
+/-- A utility function to show the pairings `(deprecatedModule, #[preferredModules])`. -/
+elab "show_deprecated_modules" : command => do
+  let directImports := deprecateModuleExt.getState (← getEnv)
+  logInfo <| "\n".intercalate <|
+    directImports.fold (init := ["Deprecated modules\n"]) fun nms (i, deps) =>
```

Comment:
I'm pretty sure there's room for bikeshedding this formatting. I think we can land something and iterate.

Reply:
Can we use the same import linking that `find_home` emits?

Reply:
I don't have time to look into this now, but I like the idea!